### PR TITLE
Simplify implementation of FoldResult and add tests

### PR DESCRIPTION
**Pull request by matthew-piziak**
@matthiasbeyer
#777

### CODE REVIEW

File: libimagutil/src/iter.rs [hunk 1]
```diff
@@ -21,29 +20,45 @@
...
+    let v: Vec<usize> = vec![1, 2, 3];
+    let mut accum: Vec<usize> = vec![];
+    let result: Result<(), &str> = v.iter().fold_result(|item| if *item == 2 {
+        Err("failure")
+    } else {
+        accum.push(*item * 2);
+        Ok(*item)
+    });
+    assert_eq!(result, Err("failure"));
+}
```
Comment (matthiasbeyer) [inline libimagutil/src/iter.rs]: Please also add an `assert!()` whether `accum` holds the values we expect it to hold.

File: libimagutil/src/iter.rs [hunk 2]
```diff
@@ -1,4 +1,3 @@
-//
```
Comment (matthiasbeyer) [inline libimagutil/src/iter.rs]: I guess this whole PR should be rebased. The file will then contain a license header...

File: libimagutil/src/iter.rs [hunk 3]
```diff
@@ -21,29 +20,45 @@
...
-              F: FnMut(Self::Item)
-        -> Result<R, E>
-    {
-        self.fold_result(R::default(), func)
-    }
-
-    /// Processes all contained items returning the last successful result or the first error.
-    /// If there are no items, returns `Ok(default)`.
-    fn fold_result<R, E, F>(self, default: R, mut func: F) -> Result<R, E>
+    /// Processes all contained items, returning the first error if it exists.
```
Comment (matthiasbeyer) [inline libimagutil/src/iter.rs]: Maybe we should also change this comment to be a bit more explanatory.

Somethink like `"Process all contained items with a function that `* -> Result<*, Error>`, breaking the iteration if the function returns an `Err(_)` and returning this error then."`

Feel free to word it your way! :smile:

File: libimagstore/src/hook/aspect.rs [hunk 4]
```diff
@@ -148,4 +148,3 @@ fn trace_hook_errors(res: HookResult<()>) -> HookResult<()> {
         }
     })
 }
-
```
Comment (matthiasbeyer) [inline libimagstore/src/hook/aspect.rs]: Please keep the newlines at the end of the files.

File: libimagstore/src/store.rs [hunk 5]
```diff
@@ -2162,4 +2162,3 @@ Hai";
     }
 
 }
-
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Please keep the newlines at the end of the files.

File: imag-bookmark/src/main.rs [hunk 6]
```diff
@@ -152,4 +152,3 @@ fn remove(rt: &Runtime) {
         .ok();
     info!("Ready");
 }
-
```
Comment (matthiasbeyer) [inline imag-bookmark/src/main.rs]: Please keep the newlines at the end of the files.

File: libimagutil/src/iter.rs (hunk 1 above)
Comment (matthiasbeyer) [inline libimagutil/src/iter.rs]: Also please add a newline at the end of the file.

Comment (matthew-piziak) [inline libimagutil/src/iter.rs]: That's strange. On my local instance I have `require-final-newline` set to true, and if I change it to false I see changes and Emacs informs me that the file does not contain a terminating newline. That implies to me that I *already* have a newline. Do you want to have your files terminated with *two* newlines (\n\n)?

File: libimagutil/src/iter.rs (hunk 2 above)
Comment (matthew-piziak) [inline libimagutil/src/iter.rs]: Done.

**Comment by matthiasbeyer** [comment]
Hi. When you rebased your changes, you put `imag-bookmark` back in the repo (probably with some merge conflict).

Care to work on this and remove the things, or, if you don't want, I can take your patches and base a new PR on them, and resolve the conflicts myself. Either way: Thanks a lot for your effort!

**Comment by matthew-piziak** [comment]
Whoops. Made a mess here. Hold on.

**Comment by matthew-piziak** [comment]
Replacing with https://github.com/matthiasbeyer/imag/pull/898

**Pull request closed without merging**